README.md
# aprico-gen
Deterministic password generator based on `scrypt` algorithm.

This javascript library generates passwords that:
- are derived deterministically purely on user inputs (a master password, a service, an ID and other options);
- are generated 100% client-side in the browser;
- do not need to store data or keep any state.

For the full implementation see [aprico.org](https://aprico.org).

Uses the [scrypt-async-js](https://github.com/dchest/scrypt-async-js) library for the `scrypt` implementation.

How it works
------------

User inputs are concatenated and hashed through the `scrypt` algorithm with a cost of `N = 2^14`. The resulting hash is then splitted, re-iterated with a lighter cost (`N = 2^5`) while arbitrary converted to a custom alphabet made of letters, numbers and/or symbols.

The result is checked for some conditions (at least a character for each alphabet, no character repeated more than 2 times) and eventually re-hashed until those conditions are met.

Install
-------

`aprico-gen` is available on NPM:

    $ npm i aprico-gen

You can use it as a CommonJS module for both node.js and the browser (For node.js there are better built-in implementations but for this specific purpose `scrypt-async-js` should be good enough):

    const aprico = require('aprico-gen');

Or directly in your browser (but without dependencies, so you may want to include also the `scrypt-async-js` library):

```html
<script src="https://unpkg.com/scrypt-async@2.0.1/scrypt-async.min.js"></script>
<script src="https://unpkg.com/aprico-gen"></script>
```

API
---

Note: As you can see, this library is written in a synchronous fashion (also `scrypt-async-js`, despite its name, is used synchronously). The library is fast enough to avoid any browser freezing. More than that, with the used parameters, from my findings this sync implementation is actually faster than an async one.

#### `getPassword(password, service, hashId, options)`

Derives a unique password and a hash based on user inputs. You can use the hash as an entropy source to implement your own alphabet or to create other deterministic derivatives. `hashId` is actually used as a salt, you can use a hash pre-generated with `getHashId` (see later) or implement your own.

Some characters are missing from the alphabet (`l`, `I`, `O` and `0`). This is by design, to avoid ambiguity in generated passwords (like `1lI0O`) and make them less prone to misreading for users.

```javascript
// default options:
let options = {
	length : 20, // min 4, max 40
	letters : true, // use letters (abcdefghijkmnopqrstuvwxyzABCDEFGHJKLMNPQRSTUVWXYZ)
	numbers : true, // use numbers (123456789)
	symbols	: true, // use symbols (_!$-+)
	variant : false // or STRING, to get a different password while preserving other inputs
}

let results = aprico.getPassword ('letmein', 'website.com', hashId, options);

console.log(results);
// {
//  	pass: 'vnscMKN_9z5a$cQC+J8G',
//  	hash: 'f9e5ce41fc655f8cdd2a103e0ed7b1db99a4618f517d6aa324e406268a650d96'
// }
```

#### `getHashId(id)`

Derives a hash from the user ID that can be used as a salt in `getPassword`. This can be saved on your end for later re-use.

```javascript
let hashId = aprico.getHashId('user@email.com');

console.log(hashId)
// "263f5a46fc0abcee64e59c086327b15cc07f0ec7b7fc16ee2ca5b791e6e63477"
```

#### `normalizeService(service)`

Returns a lowecased `service` and, if it is a URL, it is stripped down to hostname only to improve usability. Note: This function is exposed mainly as a way to reflect the normalized service in UIs as `getPassword` already calls this internally when generating a password.

```javascript
let service = aprico.normalizeService('https://www.website.com/login');

console.log(service); // 'www.website.com'
```

#### `version`

Returns the current version number.

```javascript
console.log( aprico.version ) // "1.0.0"
```

Unit Test
---------

Tests are available via npm:

    $ npm run test

Browser support
---------------

This library is written in vanilla (untraspiled) ES6, so is targeted to modern browsers only: FireFox 52+, Chrome 55+, Edge 15+, Safari 10.1+, iOS Safari 10.3+.

Changelog
---------

See [CHANGELOG.md](CHANGELOG.md)

License
-------

© Pino Ceniccola

This project is licensed under the GPLv3 License - see [LICENSE](LICENSE) for details.

CHANGELOG.md
# Changelog
All notable changes to `aprico-gen` will be documented in this file.

## 1.1.1 - 2020-08-03
### Fixed
- Updated Mocha dependency due to a security issue
- README.md updated
- Clarified in code that `script-async` is used in a synchronous way

## 1.1.0 - 2018-12-29
### Added
- Unit Tests
- README.md description
- CHANGELOG.md

## 1.0.0 - 2018-10-22
### Added
- First Public release.


test/test.js
/*!
 * aprico-gen Unit Test
 * Copyright (c) 2018 Pino Ceniccola | GPLv3
 * https://aprico.org
 */


const assert = require('assert');
const aprico = require('../aprico-gen.js');


const vectors = [
	{
		id: "",
		password: "",
		service: "",
		options: {
			length : 20,
			letters : true,
			numbers : true,
			symbols	: true,
			variant : false
		},
		normalizedService: "",
		hashId: "f9e5ce41fc655f8cdd2a103e0ed7b1db99a4618f517d6aa324e406268a650d96",
		result: "fx73Z74EXkaY+_VHk$nS"
	},
	{
		id: "user@email.com",
		password: "letmein",
		service: "MyApp",
		options: {
			length : 20,
			letters : true,
			numbers : true,
			symbols	: true,
			variant : false
		},
		normalizedService: "myapp",
		hashId: "263f5a46fc0abcee64e59c086327b15cc07f0ec7b7fc16ee2ca5b791e6e63477",
		result: "vnscMKN_9z5a$cQC+J8G"
	},
	{
		id: "user@email.com",
		password: "letmein",
		service: "MyApp",
		options: {
			length : 20,
			letters : true,
			numbers : true,
			symbols	: true,
			variant : "with variant"
		},
		normalizedService: "myapp",
		hashId: "263f5a46fc0abcee64e59c086327b15cc07f0ec7b7fc16ee2ca5b791e6e63477",
		result: "rge+JLaxyR39$NPDtbgb"
	},
	{
		id: "Another ID",
		password: "letters only",
		service: "https://some.domain.com:3000/login/page.html",
		options: {
			length : 20,
			letters : true,
			numbers : false,
			symbols	: false,
			variant : false
		},
		normalizedService: "some.domain.com:3000",
		hashId: "dafed7cc719e8d55108b5ba9625a4985fc34c138dd70a878cb8e94e79610f731",
		result: "APshNhiXrqyBcAyVfmjJ"
	},
	{
		id: "Why not Emoji? 👀",
		password: "strong password",
		service: "www.website.com/login",
		options: {
			length : 40,
			letters : true,
			numbers : true,
			symbols	: true,
			variant : false
		},
		normalizedService: "www.website.com",
		hashId: "4eb53750b4fd13c391db04e264e42670558332a48c55aac47f9ebe028a5abdba",
		result: "n!7a55Ga6bcH58en$K-mYd!!GD+h9LNpp+Ey$76E"
	},
	{
		id: "Very long id: His talent was as natural as the pattern that was made by the dust on a butterfly's wings. At one time he understood it no more than the butterfly did and he did not know when it was brushed or marred.",
		password: "5-digit PIN",
		service: "CustomService",
		options: {
			length : 5,
			letters : false,
			numbers : true,
			symbols	: false,
			variant : false
		},
		normalizedService: "customservice",
		hashId: "dabfadb8520288cf28cda3e95b3b16c30c33d0fac92a1e0e0dbb2773741c5f43",
		result: "16549"
	},
	{
		id: "Another Test",
		password: "with emoji 🙃",
		service: "http://intranet.lan/login",
		options: {
			length : 10,
			letters : true,
			numbers : true,
			symbols	: true,
			variant : false
		},
		normalizedService: "intranet.lan",
		hashId: "adac3e4cca4aa5b3628192973d1f138714aada28590f919b75a06b3317366d1e",
		result: "mbK$ity8$L"
	},
	{
		id: "Last one.",
		password: "very long password: His talent was as natural as the pattern that was made by the dust on a butterfly's wings. At one time he understood it no more than the butterfly did and he did not know when it was brushed or marred.",
		service: "https://accounts.google.com/signin/v2/identifier?continue=https%3A%2F%2Fmail.google.com%2Fmail%2F&service=mail",
		options: {
			length : 20,
			letters : false,
			numbers : true,
			symbols	: true,
			variant : false
		},
		normalizedService: "accounts.google.com",
		hashId: "e3a29d04caeedd882d407528488b0d597fa4192ccbe1d6948eafc7b908465905",
		result: "8-!1-!4!5!__426$$278"
	}
]



describe("aprico-gen Unit Tests", () => {

	it("should normalize services", (done) => {
		vectors.forEach(v => {
			let service = aprico.normalizeService(v.service);
			assert.deepEqual(v.normalizedService, service);
		});
		done();
	});

	it("should generate correct Hash IDs", (done) => {
		vectors.forEach(v => {
			let hashId = aprico.getHashId(v.id);
			assert.deepEqual(v.hashId, hashId);
		});
		done();
	});

	it("should generate correct passwords", (done) => {
		vectors.forEach(v => {
			let results = aprico.getPassword(v.password, v.normalizedService, v.hashId, v.options);
			assert.deepEqual(v.result, results.pass);
		});
		done();
	});

});



aprico-gen.js
/*!
 * aprico-gen
 * Deterministic password generator library based on scrypt algorithm.
 * Copyright (c) 2018 Pino Ceniccola | GPLv3
 * https://aprico.org
 */

const aprico = (()=> {
	'use strict';

	const scrypt = (typeof module !== 'undefined' && module.exports) ? require('scrypt-async') : window.scrypt;

	if (typeof scrypt !== 'function')
		throw new Error("aprico requires scrypt-async-js library.");

	const VERSION = "1.1.1";

	const SCRYPT_COST = {
    	N: Math.pow(2,14),
    	r: 8,
    	p: 1,
    	dkLen: 32,
    	encoding: 'hex'
	};

	const SCRYPT_COST_FAST = {
    	N: Math.pow(2,5),
    	r: 8,
    	p: 1,
    	dkLen: 32,
    	encoding: 'hex'
	};

	const ALPHABET = {
		letters : 'abcdefghijkmnopqrstuvwxyzABCDEFGHJKLMNPQRSTUVWXYZ',
		numbers : '123456789',
		symbols : '_!$-+'
	};

	let options = {
		length : 20,
		letters : true,
		numbers : true,
		symbols	: true,
		variant : false
	};


	/**
	* Arbitrary conversion from one character set to another,
	* using Base 10 as an intermediate base for conversion.
	* Used by _hashToAlphabet().
	* (Adapted from: https://rot47.net/base.html)
	* @param {string} src - String to convert.
	* @param {string} srctable - Source character set.
	* @param {string} srcdest - Destination character set.
	* @returns {string} Converted string.
	*/
	const _convert = (src, srctable, desttable) => {
		const srclen = srctable.length;
		const destlen = desttable.length;
		// Convert to base 10
		let val = 0;
		for (let i=0, numlen = src.length; i<numlen; i++) {
			val = val * srclen + srctable.indexOf(src.charAt(i));
		};
		if (val<0) {return 0;}
		// Then convert to destination base
		let r = val % destlen;
		let res = desttable.charAt(r);
		let q = Math.floor(val/destlen);
		while (q) {
			r = q % destlen;
			q = Math.floor(q/destlen);
			res = desttable.charAt(r) + res;
		}
		return res;
	};


	/**
	 * Convert full hex hash to password format
	 * set by options (alphabet and length)
	 * @param {string} hash - Hash in hex format.
	 * @returns {string} The generated password.
	 */
	const _hashToAlphabet = (hash) => {
		let alphabet = '';
		if (options.symbols) alphabet += ALPHABET.symbols;
		if (options.numbers) alphabet += ALPHABET.numbers;
		if (options.letters) alphabet += ALPHABET.letters;

		let result = '';

		// We split the full hash because... 32bit
		let split = hash.match(/(.{1,7})/g);
		for (let i = 0, l = split.length-1; i < l; i++) {

			//[debug] console.log('Split', split);

			// Re-hash every slice when revealing a potentially
			// large chunk of the original hash.
			if (options.length > 9) {

				//[debug] console.log('Re-hashing', hash, split[l-i-1]);

				scrypt (hash, split[l-i-1], SCRYPT_COST_FAST, (hash) => split = hash.match(/(.{1,7})/g) );
			}

			result += _convert(split[i], '0123456789abcdef', alphabet);

		};

		//[debug] console.log('Full converted string', result);

		// Better safe than sorry
		if (result.length<=options.length) return ".";

		// Trim password to options.length
		let offset = ((result.length-options.length)/2)|0;

		//[debug] console.log('Trim offset', offset);

		result = result.substring(offset, offset + options.length);

		// Make sure first character is [A-z]
		// because... silly password rules
		let firstChar = result.charAt(0);
		if (options.letters && options.length > 6 && !/[a-zA-Z]/.test(firstChar)) {

			//[debug] console.log('First character to be replaced', result);

			result = result.replace(firstChar, _convert(firstChar, ALPHABET.symbols+ALPHABET.numbers, ALPHABET.letters));
		}

		return result;
	};


	/**
	 * Check if the generated password has characters from
	 * every alphabet set. If not, re-hash until conditions are met.
	 * @param {object} results - Password + Hash.
	 * @returns {object} Password + Hash.
	 */
	const _checkAndIterate = (results) => {

		let success = false;

		while (!success) {

			success = true;

			// Check for...
			// 1. at least one uppercase, one lowecase letter, no letter repeated three times (eg. aaa)
			if ( options.letters && ( !/[a-z]/.test(results.pass) || !/[A-Z]/.test(results.pass) || /([A-z])\1{2}/.test(results.pass) ) ) success = false;

			// 2. at least one number and no number repeated three times (eg. 111)
			if ( success && options.numbers && ( !/[\d]/.test(results.pass) || /([\d])\1{2}/.test(results.pass) ) ) success = false;

			// 3. at least one symbol and no symbol repeated three times (eg. $$$)
			if ( success && options.symbols && ( !/[!$\-+_]/.test(results.pass) || /([!$\-+_])\1{2}/.test(results.pass) ) ) success = false;


			if (!success) {

				//[debug] console.log('Iterating', results);

				// Re-hash until conditions are met.
				scrypt (results.hash, results.pass, SCRYPT_COST_FAST, (hash) => results.hash = hash );

				results.pass = _hashToAlphabet(results.hash);

			}

		}

		return results;

	};


	/**
	 * Merge and check user options.
	 * @param {object} user_options	- User options.
	 */
	const _checkOptions = (user_options) => {

		for (let opt in options)
			if (user_options.hasOwnProperty(opt))
				options[opt] = user_options[opt];

		options.letters = !!options.letters;
		options.numbers = !!options.numbers;
		options.symbols = !!options.symbols;

		if (!options.letters && !options.symbols && !options.numbers)
			throw new Error("At least one character set (letters, numbers, symbols) must be chosen.");

		options.length = +options.length;

		if (typeof options.length !== 'number' || options.length < 4 || options.length > 40)
			throw new Error("Password length must be a number between 4 and 40.");

	};


	/**
	 * Deterministic password generation main function.
	 * @param {string} password	- User Master Password.
	 * @param {string} service - User service/domain.
	 * @param {string} hashId - Precomputed hash ID.
	 * @param {object} user_options	- User options.
	 * @returns {object} Password + Hash.
	 */
	const getPassword = (password, service, hashId, user_options) => {

		if (user_options) _checkOptions(user_options);

		service = normalizeService(service);

		let pass = password + '.' + service + '.' + options.length + (+options.letters) + (+options.symbols) + (+options.numbers);

		if (options.variant && typeof options.variant === 'string') pass += '.' + options.variant;

		//[debug] console.log('String to hash', pass);

		let results = {};

		// Note: scrypt callback is called immediately, in a synchronous fashion here.
		// See https://github.com/dchest/scrypt-async-js#usage
		scrypt (pass, hashId, SCRYPT_COST, (hash) => results.hash = hash);

		results.pass = _hashToAlphabet(results.hash);
		results = _checkAndIterate(results);

		//[debug] console.log('Results', results);

		return results;

	};



	/**
	 * Generate a hash from the user ID.
	 * @param {string} id - User ID.
	 * @returns {string} Hash ID.
	 */
	const getHashId = (id) => {
		let output = '';

		// In order to create a hash from the ID using scrypt,
		// we need to generate some deterministic salt and it's
		// not a bad thing.
		// Rationale: ID is a salt. It's not a secret.
		// We are not hashing a password.
		// We are converting ID to a hash more for convenience
		// than security here.
		let salt = Math.pow(id.length, (id.match(/[aeiou]/gi) || [0,0,0]).length)+'';
		salt = _convert(salt, '0123456789.e+Infity', ALPHABET.numbers+ALPHABET.symbols+ALPHABET.letters)+'';

		//[debug] console.log('Hash ID salt',salt);

		scrypt(id, salt, SCRYPT_COST, (hash) => output = hash);

		return output;
	};


	/**
	 * If Service is a URL, it is stripped down to hostname (and
	 * perhaps port number) to improve usability.
	 * @param {string} service - User Service.
	 * @returns {string} - Normalized Service.
	 */
	const normalizeService = (service) => {

		service = service.trim().toLowerCase();

		// Strip http(s)://
		if (service.substring(0, 4) == 'http') {
			service = service.substring(service.indexOf('://')+3, service.length);
		}

		// if string contains any "/" take only the first part
		if (service.indexOf('.') !== -1 && service.indexOf('/') !== -1) {
			service = service.split('/');
			service = service[0];
		}

		//[debug] console.log('Normalized Service', service);

		return service;
	};


	return {
		getPassword : getPassword,
		getHashId : getHashId,
		normalizeService : normalizeService,
		version : VERSION
	};

})();


if (typeof module !== 'undefined' && module.exports) {
	module.exports = aprico;
} else {
	window.aprico = aprico;
}